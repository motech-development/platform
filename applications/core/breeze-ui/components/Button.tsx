import { FC, HTMLAttributes } from 'react';

type TColour = 'danger' | 'primary' | 'secondary' | 'success';

type TSize = 'sm' | 'md' | 'lg';

const styles = (colour: TColour, size: TSize, block: boolean) => {
  let colourStyles: string;
  let sizeStyles: string;

  switch (colour) {
    case 'danger':
      colourStyles =
        'bg-red-600 hover:bg-red-700 focus:ring-red-500 text-gray-100';
      break;
    case 'secondary':
      break;
    case 'success':
      colourStyles =
        'bg-green-600 hover:bg-green-700 focus:ring-green-500 text-gray-100';
      break;
    default:
      colourStyles =
        'bg-blue-600 hover:bg-blue-700 focus:ring-blue-500 text-gray-100';
  }

  switch (size) {
    case 'sm':
      break;
    case 'lg':
      break;
    default:
      sizeStyles = 'py-2 px-4';
  }

  const blockStyles = block ? 'w-full' : '';

  return `${colourStyles} ${blockStyles} ${sizeStyles}`;
};

interface IButtonProps extends HTMLAttributes<HTMLButtonElement> {
  block?: boolean;
  colour?: TColour;
  size?: TSize;
  disabled?: boolean;
  loading?: boolean;
  type?: 'submit' | 'reset' | 'button';
}

const Button: FC<IButtonProps> = ({
  block = false,
  colour = 'primary',
  size = 'md',
  type = 'button',
  ...rest
}) => (
  <button
    className={`group relative shadow-2xl flex font-display font-semibold justify-center text-lg font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 ${styles(
      colour,
      size,
      block,
    )}`}
    type={type}
    {...rest}
  ></button>
);

export default Button;
