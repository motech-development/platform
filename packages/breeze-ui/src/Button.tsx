import React, { FC } from 'react';
import styled from 'styled-components';

const BaseButton = styled.button`
  appearance: none;
  vertical-align: middle;
  user-select: none;
  white-space: nowrap;
`;

export interface IButtonProps {
  type?: 'submit' | 'reset' | 'button';
}

const Button: FC<IButtonProps> = ({ children, type = 'button' }) => {
  return <BaseButton type={type}>{children}</BaseButton>;
};

export default Button;
