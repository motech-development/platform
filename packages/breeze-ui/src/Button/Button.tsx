import React, { FC, HTMLAttributes, memo } from 'react';
import styled, { ThemeProvider } from 'styled-components';
import BaseButton, {
  buttonTheme,
  IBaseButtonProps,
} from '../BaseButton/BaseButton';
import Loader, { ILoaderProps } from '../Loader/Loader';

const loaderSizes = {
  lg: {
    height: '36px',
    margin: '-18px 0 0 -18px',
    width: '36px',
  },
  md: {
    height: '28px',
    margin: '-14px 0 0 -14px',
    width: '28px',
  },
  sm: {
    height: '20px',
    margin: '-10px 0 0 -10px',
    width: '20px',
  },
};

interface IButtonLoaderProps extends ILoaderProps {
  $size: keyof typeof loaderSizes;
}

const ButtonLoader = styled(Loader)<IButtonLoaderProps>`
  ${({ $size }) => `
    height: ${loaderSizes[$size].height};
    margin: ${loaderSizes[$size].margin};
    width: ${loaderSizes[$size].width};
  `}
`;

interface IButtonContainerProps {
  $isLoading: boolean;
}

const ButtonContainer = styled.span<IButtonContainerProps>`
  ${({ $isLoading }) => `
    visibility: ${$isLoading ? 'hidden' : 'visible'};
  `}
`;

export interface IButtonProps
  extends IBaseButtonProps,
    HTMLAttributes<HTMLButtonElement> {
  block?: boolean;
  disabled?: boolean;
  loading?: boolean;
  type?: 'submit' | 'reset' | 'button';
}

const Button: FC<IButtonProps> = ({
  block = false,
  children,
  colour = 'primary',
  disabled = false,
  loading = false,
  size = 'md',
  type = 'button',
  ...rest
}) => {
  const loadingColour = colour === 'secondary' ? 'default' : 'secondary';

  return (
    <ThemeProvider theme={buttonTheme}>
      <BaseButton
        block={block}
        colour={colour}
        type={type}
        size={size}
        disabled={disabled || loading}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...rest}
      >
        {loading && <ButtonLoader colour={loadingColour} $size={size} />}

        <ButtonContainer $isLoading={loading}>{children}</ButtonContainer>
      </BaseButton>
    </ThemeProvider>
  );
};

export default memo(Button);
