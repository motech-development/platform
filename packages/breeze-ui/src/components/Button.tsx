import { ButtonHTMLAttributes, FC, forwardRef } from 'react';
import { classNames } from '../utils/className';
import TSize from '../utils/size';
import TTheme from '../utils/theme';
import BaseButton from './BaseButton';
import Loader from './Loader';

const loaderSize = (size: TSize) => {
  switch (size) {
    case 'lg':
      return 'w-7 h-7 -mt-3.5 -ml-3.5';
    case 'sm':
      return 'w-3 h-3 -mt-1.5 -ml-1.5';
    default:
      return 'w-5 h-5 -mt-2.5 -ml-2.5';
  }
};

type TButtonType = 'submit' | 'reset' | 'button';

export interface IButtonProps
  extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'size'> {
  block?: boolean;
  colour?: TTheme;
  disabled?: boolean;
  loading?: boolean;
  size?: TSize;
  type?: TButtonType;
}

const Button: FC<IButtonProps> = forwardRef<HTMLImageElement, IButtonProps>(
  (
    {
      block = false,
      children,
      colour = 'primary',
      disabled = false,
      loading = false,
      size = 'md',
      type = 'button',
      ...rest
    },
    ref,
  ) => (
    <BaseButton
      as="button"
      block={block}
      colour={colour}
      disabled={disabled || loading}
      ref={ref}
      size={size}
      type={type}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    >
      {loading && (
        <Loader
          className={classNames('absolute top-2/4 left-2/4', loaderSize(size))}
        />
      )}

      <span className={classNames(loading ? 'invisible' : '')}>{children}</span>
    </BaseButton>
  ),
);

export default Button;
