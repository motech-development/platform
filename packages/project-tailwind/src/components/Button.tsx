import { ElementType } from 'react';
import { Box, PolymorphicComponentProps } from 'react-polymorphic-box';
import {
  Sizing,
  Themes,
  TSizing,
  TTheme,
  useTailwind,
} from '../utilities/tailwind';

/** Default component element */
const DEFAULT_ELEMENT = 'button';

/** Button component props */
export type TButtonProps<E extends ElementType> = PolymorphicComponentProps<
  E,
  {
    /** Renders component as a block element */
    block?: boolean;

    /** Component theme */
    theme?: TTheme;

    /** Component size */
    size?: TSizing;
  }
>;

/**
 * Flexible call to action button
 *
 * @param props - Button component props
 *
 * @returns Button component
 */
export function Button<E extends ElementType = typeof DEFAULT_ELEMENT>({
  block = false,
  className,
  disabled,
  theme = Themes.PRIMARY,
  size = Sizing.MD,
  ...rest
}: TButtonProps<E>) {
  const { createStyles } = useTailwind(theme, size);

  const buttonStyles = createStyles({
    classNames: [
      {
        'cursor-not-allowed bg-opacity-75 hover:bg-opacity-75': disabled,
        'w-full': block,
      },
      'group items-center relative shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none focus:outline-none focus:ring-2 focus:ring-offset-2',
      className,
    ],
    sizing: {
      lg: ['py-4 px-6 text-lg'],
      md: ['py-2 px-4 text-base'],
      none: ['p-0 text-sm'],
      sm: ['py-0 px-2 text-sm'],
    },
    theme: {
      danger: [
        'bg-red-600 text-white',
        {
          'hover:bg-red-700 focus:ring-red-500': !disabled,
        },
      ],
      primary: [
        'bg-blue-600 text-white',
        {
          'hover:bg-blue-700 focus:ring-blue-500': !disabled,
        },
      ],
      secondary: [
        'bg-gray-100 text-gray-600',
        {
          'hover:bg-gray-300 focus:ring-gray-500': !disabled,
        },
      ],
      success: [
        'bg-green-600 text-white',
        {
          'hover:bg-green-700 focus:ring-green-500': !disabled,
        },
      ],
      warning: [
        'bg-yellow-600 text-white',
        {
          'hover:bg-yellow-700 focus:ring-yellow-500': !disabled,
        },
      ],
    },
  });

  return (
    <Box
      className={buttonStyles}
      as={DEFAULT_ELEMENT}
      {...{ disabled }}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    >
      Test
    </Box>
  );
}
