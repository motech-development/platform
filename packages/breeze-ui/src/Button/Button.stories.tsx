import { boolean, select, text, withKnobs } from '@storybook/addon-knobs';
import BaseStyles from '../BaseStyles/BaseStyles';
import Button from './Button';

export default {
  component: Button,
  decorators: [withKnobs],
};

const colour = {
  Danger: 'danger',
  Primary: 'primary',
  Secondary: 'secondary',
  Success: 'success',
};
const size = {
  Large: 'lg',
  Medium: 'md',
  Small: 'sm',
};

export const BasicButton = {
  name: 'Basic button',
  render: () => (
    <>
      <BaseStyles />
      <Button
        block={boolean('Block display', false)}
        colour={
          select('Colour', colour, 'primary') as
            | 'danger'
            | 'primary'
            | 'secondary'
            | 'success'
        }
        size={select('Size', size, 'md') as 'sm' | 'md' | 'lg'}
        loading={boolean('Loading', false)}
      >
        {text('Title', 'Button')}
      </Button>
    </>
  ),
};
