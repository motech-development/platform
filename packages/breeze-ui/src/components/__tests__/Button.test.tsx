import { render } from '@testing-library/react';
import Button from '../Button';

describe('Button', () => {
  describe('size', () => {
    it('should render when size is sm', () => {
      const { container } = render(<Button size="sm">Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render when size is md', () => {
      const { container } = render(<Button>Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render when size is lg', () => {
      const { container } = render(<Button size="lg">Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('theme', () => {
    it('should have the correct danger styles', () => {
      const { container } = render(<Button colour="danger">Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct primary styles', () => {
      const { container } = render(<Button colour="primary">Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct secondary styles', () => {
      const { container } = render(<Button colour="secondary">Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct success styles', () => {
      const { container } = render(<Button colour="success">Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct warning styles', () => {
      const { container } = render(<Button colour="warning">Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('block', () => {
    it('should not be displayed as a block element by default', () => {
      const { container } = render(<Button>Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should not be displayed as a block element', () => {
      const { container } = render(<Button block={false}>Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should be displayed as an block element', () => {
      const { container } = render(<Button block>Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when sm', () => {
      const { container } = render(
        <Button block size="sm">
          Hello
        </Button>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when md', () => {
      const { container } = render(
        <Button block size="md">
          Hello
        </Button>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when lg', () => {
      const { container } = render(
        <Button block size="lg">
          Hello
        </Button>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('loading', () => {
    it('should render correctly when loading', () => {
      const { container } = render(<Button loading>Hello</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
