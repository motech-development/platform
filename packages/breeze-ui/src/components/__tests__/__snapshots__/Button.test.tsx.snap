// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button block should be displayed as an block element 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button block should have the correct padding when lg 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-4 px-6 text-lg w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button block should have the correct padding when md 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button block should have the correct padding when sm 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-0 px-2 text-sm w-full hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button block should not be displayed as a block element 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button block should not be displayed as a block element by default 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button loading should render correctly when loading 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base cursor-not-allowed bg-opacity-75 hover:bg-opacity-75"
  type="button"
>
  <svg
    class="animate-spin absolute top-2/4 left-2/4 w-5 h-5 -mt-2.5 -ml-2.5"
    fill="none"
    viewBox="0 0 24 24"
  >
    <circle
      class="opacity-25"
      cx="12"
      cy="12"
      r="10"
      stroke="currentColor"
      stroke-width="4"
    />
    <path
      class="opacity-50"
      d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
      fill="currentColor"
    />
  </svg>
  <span
    class="invisible"
  >
    Hello
  </span>
</button>
`;

exports[`Button size should render when size is lg 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-4 px-6 text-lg hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button size should render when size is md 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button size should render when size is sm 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-0 px-2 text-sm hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button theme should have the correct danger styles 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-red-600 text-white py-2 px-4 text-base hover:bg-red-700 focus:ring-red-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button theme should have the correct primary styles 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-blue-600 text-white py-2 px-4 text-base hover:bg-blue-700 focus:ring-blue-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button theme should have the correct secondary styles 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-gray-100 hover:bg-gray-300 focus:ring-gray-500 text-gray-600 py-2 px-4 text-base hover:bg-gray-700 focus:ring-gray-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button theme should have the correct success styles 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-green-600 text-white py-2 px-4 text-base hover:bg-green-700 focus:ring-green-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;

exports[`Button theme should have the correct warning styles 1`] = `
<button
  class="group items-center relative mt-auto shadow-2xl inline-flex font-display justify-center font-medium transition-colors select-none bg-yellow-600 text-white py-2 px-4 text-base hover:bg-yellow-700 focus:ring-yellow-500 focus:outline-none focus:ring-2 focus:ring-offset-2"
  type="button"
>
  <span
    class=""
  >
    Hello
  </span>
</button>
`;
